### PR TITLE
Adapt max_angle calculation to match last angle step

### PR DESCRIPTION
**Pull request by cdoehn**
## Description

In this PR two things are changed:

1. The definition of max_angle in the published laserscan and the parameter angle_end is changed. Now they represent the corresponding angle value of the last measurement. Before this PR the max_angle and end_angle itself are not included in the measurements.

2. The scanner range with 0.1 degree resolution is not symmetrical. The range is -137.5...137.4 degrees. Therefore we change the default range to a symmetrical range of angle_start=-137.4..angle_end=137.4 degrees. (With the new meaning of angle_end, that means the corresponding angle of the last measurement is 137.4 degree)

### Things to add, update or check by the maintainers before this PR can be merged.

* [x] Public api function documentation
* [x] ~Architecture documentation reflects actual code structure~
* [x] ~[Tutorials](https://wiki.ros.org/pilz_robots/Tutorials/)~
* [x] ~Overview on [ROS wiki](https://wiki.ros.org/pilz_robots)~
* [x] Package Readme ([example pilz_robots](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/README.md))
* [x] Good commit messages ([some tips](https://dev.to/jacobherrington/how-to-write-useful-commit-messages-my-commit-message-template-20n9))
* [x] CHANGELOG.rst updated
* [x] Copyright headers
* [x] ~Examples~

### Review Checklist
* [x] ~Soft- and hardware architecture (diagrams and description)~
* [x] Test review (test plan and individual test cases)
* [x] Documentation describes purpose of file(s) and responsibilities
* [x] ~Code (coding rules, style guide)~

### Release planning (please answer)
* [x] When is the new feature released?
* [x] ~Which dependent packages do have to be released simultaneously?~

### Hardware tests
- [x] Perform hardware tests

### CODE REVIEW

**Comment by cdoehn** [comment]
The behavior for some examples of angular ranges is now:

<pre>roslaunch---&gt;           default         -1deg..1deg     -1.1deg...0.9deg        -1.16...0.9deg          -137.5deg...-87.5deg

start request
 angle_start            0               1365            1364                    1363                    0
 angle_end              2750            1387            1386                    1384                    502

laser scan message
 angle_min              -137.5          -1.0            -1.1                    -1.2                    -137.5
 angle_max              137.3           1.0             0.9                     1.0                     -87.5
 angle_increment        0.2             0.2             0.2                     0.2                     0.2
 number of values       1375            11              11                      12                      251</pre>

**Comment by agutenkunst** [comment]
Please add a few words to the description above and state the intendend behaviour before/after this PR

**Comment by jschleicher** [comment]
Thanks for the comparison table! That looks great. The only concern I have is the default: The range should be symmetrical. Isn't it possible to get values from -137.5..+137.5 degrees? Otherwise we'd have to lower the default range to (-137.4,+137.4), which would include a value at the center (zero degrees) as well ;-)
How would the default change, if switching to 0.1 degrees resolution?
@ct2034 Could you please take a look and review the technical details of the PR? Thank you!

File: standalone/include/psen_scan_v2_standalone/data_conversion_layer/laserscan_conversions.h [hunk 1]
```diff
@@ -115,7 +115,7 @@ LaserScanConverter::calculateMaxAngle(const std::vector<data_conversion_layer::m
       std::accumulate(frames.begin(), frames.end(), uint16_t{ 0 }, [](uint16_t total, const auto& frame) {
         return total + frame.measurements().size();
       });
-  return min_angle + resolution * static_cast<int>(number_of_samples);
+  return min_angle + resolution * static_cast<int>(number_of_samples - 1);
```
Comment (ct2034) [inline standalone/include/psen_scan_v2_standalone/data_conversion_layer/laserscan_conversions.h:118]: it seem like you are changing it at two places now. You change it at the sending of the request and here. why?

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp [hunk 2]
```diff
@@ -85,7 +85,7 @@ TEST(LaserScanConversionsTest, laserScanShouldContainCorrectMinMaxScanAngleAfter
   ASSERT_NO_THROW(scan_ptr.reset(new LaserScan{ data_conversion_layer::LaserScanConverter::toLaserScan({ frame }) }););
 
   const util::TenthOfDegree expected_max_scan_angle{ frame.fromTheta() +
-                                                     frame.resolution() * (int)frame.measurements().size() };
+                                                     frame.resolution() * ((int)frame.measurements().size() - 1) };
```
Comment (ct2034) [inline standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:88]: and here again. maybe I am misunderstanding something but this seems off

**Review by ct2034 — CHANGES_REQUESTED** [review]
* I can genrerally say that this works well with the scanner. the slam_toolbox does not complain anymore about the off-by-one error. 
* But I am still not sure about the choice of `angle_start` and `angle_end params`.
* To clarify: based on the data by Karl, I can say that with 0.2 resolution we have no measurement at 0° (only at 0.1 and -0.1) but with 0.1 resolution there is a measurement at 0°. I think we will maybe hav to take that into account, too unfortunately.

File: launch/psen_scan_v2.launch [hunk 3]
```diff
@@ -35,7 +35,7 @@ along with this program.  If not, see <https://www.gnu.org/licenses/>.
   <arg name="angle_start" default="$(eval -137.5 /180.0 * pi)" />
 
   <!-- End angle of measurement in radian -->
-  <arg name="angle_end" default="$(eval 137.5 /180.0 * pi)" />
+  <arg name="angle_end" default="$(eval 137.3 /180.0 * pi)" />
```
[the remark this answers is not in the document]
Comment (cdoehn) [reply launch/psen_scan_v2.launch:38]: Only the meaning of the parameter changed

[the remark this answers is not in the document]
Comment (cdoehn) [reply launch/psen_scan_v2.launch:38]: > To clarify: based on the data by Karl, I can say that with 0.2 resolution we have no measurement at 0° (only at 0.1 and -0.1) but with 0.1 resolution there is a measurement at 0°. I think we will maybe hav to take that into account, too unfortunately.

With angle_start:=-137.4 you should have a value at 0° also for resolution:=0.2

File: standalone/include/psen_scan_v2_standalone/data_conversion_layer/laserscan_conversions.h (hunk 1 above)
Comment (cdoehn) [reply standalone/include/psen_scan_v2_standalone/data_conversion_layer/laserscan_conversions.h:118]: This position in the code only affects the calculation of the laser scan message and has nothing to do with the start request

**Comment by ct2034** [comment]
I am sorry, but I have some more issues with the implementation:
1. when i change the start and end angles to any other value, I get no data from the scanner. Maybe I am doing something wrong but it is really weird.
1. when I use the new values from the launchfile (which I dont like) and I analyze the data, I get: 

```
[WARN] [1620307835.687052]: param angle_start:
 -137.5
[WARN] [1620307835.688189]: param angle_end:
 137.3
[WARN] [1620307835.689148]: msg angle_min:
 -137.49999984200906
[WARN] [1620307835.690072]: msg angle_max:
 137.29999824273276
[WARN] [1620307835.690915]: msg angle_increment:
 0.19999999845072666
[WARN] [1620307835.691750]: msg angle_increment * len(ranges):
 274.99999786974917
[WARN] [1620307835.692645]: msg angle_min + angle_increment * len(ranges):
 137.49999802774008
```

It is interesting, how the min and max angles from the message can match of with the (samller) configured range while the scanner still seems to send the full 275 degrees. something is very wrong here.

File: test/hw_tests/hwtest_scan_range.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import sys
+import unittest
+from math import degrees
+
+import rospy
+from sensor_msgs.msg import LaserScan
+
+import pytest
+
+PLACES = 6
```
Comment (agutenkunst) [inline test/hw_tests/hwtest_scan_range.py:12]: Please choose a more expressive variable name

File: test/hw_tests/hwtest_scan_range.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class HwtestScanRange(unittest.TestCase):
```
Comment (agutenkunst) [inline test/hw_tests/hwtest_scan_range.py:15]: Please add a short description what this test intents todo

File: test/hw_tests/hwtest_scan_range.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        rospy.loginfo(f"msg angle_min:\n {degrees(message.angle_min)}")
+        rospy.loginfo(f"msg angle_max:\n {degrees(message.angle_max)}")
+        rospy.loginfo(
+            f"msg angle_increment:\n {degrees(message.angle_increment)}")
+        rospy.loginfo("msg angle_increment * len(ranges):\n" +
+                      f" {degrees(message.angle_increment * len(message.ranges))}")
+        rospy.loginfo("msg angle_increment * (len(ranges)-1):\n" +
+                      f" {degrees(message.angle_increment * (len(message.ranges)-1))}")
+        rospy.loginfo("msg angle_min + angle_increment * (len(ranges)-1):\n" +
+                      f" {degrees(message.angle_min + message.angle_increment * (len(message.ranges)-1))}")
```
Comment (agutenkunst) [inline test/hw_tests/hwtest_scan_range.py:43]: This should not be needed. Rather print a verbose output if the test fails.

**Comment by martiniil** [comment]
> My suggestion for always getting a measuremant at 0.0° and to have a symmetric range would be:
> 
>     * have -137.5 / 137.5 as default range
> 
>     * round the range down to nearest multiple of the resoultion (here  -137.4 / 137.4)
> 
>     * add information output to the user about this rounding.
> 
>     * This would have the effect, that for a resolution of 0.2°, the start angle from the parameters would only match the min angle in the message if their tenth deegrees is even. Which, I would say, is not to big of an assumption.

So you would always centralize the data with respect to the given range? Why not.
What is the point of always having a measurement at 0.0?

**Comment by martiniil** [comment]
integrationtest_scanner_api fails on windows: https://github.com/PilzDE/psen_scan_v2/pull/180/checks?check_run_id=2818190149

**Comment by agutenkunst** [comment]
With https://github.com/PilzDE/psen_scan_v2/pull/180/commits/e0158e12b325c17d40290096c956e9071d4b6230
and real hardware `rostest psen_scan_v2 hwtest_scan_range.test config:=big_resolution.yaml` results in

```
... logging to /home/alex/.ros/log/rostest-rechenknecht-64727.log
[ROSUNIT] Outputting test results to /home/alex/.ros/test_results/psen_scan_v2/rostest-test_hw_tests_hwtest_scan_range.xml
[ INFO] [1623741967.251400858]: Start scanner called.
[ INFO] [1623741967.252944505]: No host ip set! Using local ip: 192.168.0.122
[ INFO] [1623741967.255726413]: Scanner started successfully.
[ INFO] [1623741967.313849120]: Publishing laser scan with angle_min=-137.4 angle_max=136.6 angle_increment=1.0 degrees. 275 angle values.
[ INFO] [1623741971.251388195]: Stop scanner called.
[ INFO] [1623741971.252316667]: Scanner stopped successfully.
[Testcase: testscan_range] ... ok

[ROSTEST]-----------------------------------------------------------------------

[psen_scan_v2.rosunit-scan_range/test_one_message_received][FAILURE]------------
2.398082392240209 != 2.3841196186840534 within 6 places (0.013962773556155383 difference) : angle_max of the laserscan message is 2.3841196186840534 but should be 2.398082392240209.
  File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/alex/catkin_ws_psen_scan_v2_only/src/psen_scan_v2/test/hw_tests/hwtest_scan_range.py", line 40, in test_one_message_received
    self.assertAlmostEqual(self.angle_end, angle_max, DECIMAL_PLACE_ACCURACY,
  File "/usr/lib/python3.8/unittest/case.py", line 966, in assertAlmostEqual
    raise self.failureException(msg)
--------------------------------------------------------------------------------


SUMMARY
 * RESULT: FAIL
 * TESTS: 1
 * ERRORS: 0
 * FAILURES: 1

rostest log file is in /home/alex/.ros/log/rostest-rechenknecht-64727.log
```

[e0158e1.zip](https://github.com/PilzDE/psen_scan_v2/files/6653662/e0158e1.zip)

**Review by ct2034 — APPROVED** [review]

**Pull request merged**